libr/socket/r2pipe.c

```diff
@@ -233,93 +234,101 @@ R_API R2Pipe *r2pipe_open_dl(const char *libr_path) {
 }
 
 R_API R2Pipe *r2pipe_open(const char *cmd) {
-	R2Pipe *r2pipe = r2pipe_new ();
-	if (!r2pipe) {
+	R2Pipe *r2p = r2pipe_new ();
+	if (!r2p) {
 		return NULL;
 	}
-	if (!cmd) {
-		r2pipe->child = -1;
-		return r2pipe_open_spawn (r2pipe);
+	// r2p->magic = R2P_MAGIC;
+	if (R_STR_ISEMPTY (cmd)) {
+		r2p->child = -1;
+		return r2p_open_spawn (r2p, cmd);
 	}
 #if __WINDOWS__
-	w32_createPipe (r2pipe, cmd);
-	r2pipe->child = (int)(r2pipe->pipe);
+	w32_createPipe (r2p, cmd);
+	r2p->child = (int)(r2p->pipe);
 #else
-	int r = pipe (r2pipe->input);
+	int r = pipe (r2p->input);
 	if (r != 0) {
 		eprintf ("pipe failed on input\n");
-		r2pipe_close (r2pipe);
+		r2pipe_close (r2p);
 		return NULL;
 	}
-	r = pipe (r2pipe->output);
+	r = pipe (r2p->output);
 	if (r != 0) {
 		eprintf ("pipe failed on output\n");
-		r2pipe_close (r2pipe);
+		r2pipe_close (r2p);
 		return NULL;
 	}
 #if LIBC_HAVE_FORK
-	r2pipe->child = fork ();
+	r2p->child = fork ();
 #else
-	r2pipe->child = -1;
+	r2p->child = -1;
 #endif
-	if (r2pipe->child == -1) {
-		r2pipe_close (r2pipe);
+	if (r2p->child == -1) {
+		r2pipe_close (r2p);
 		return NULL;
 	}
-	env ("R2PIPE_IN", r2pipe->input[0]);
-	env ("R2PIPE_OUT", r2pipe->output[1]);
+	env ("R2PIPE_IN", r2p->input[0]);
+	env ("R2PIPE_OUT", r2p->output[1]);
 
-	if (r2pipe->child) {
-		char ch = 1;
+	if (r2p->child) {
+		char ch = -1;
 		// eprintf ("[+] r2pipeipe child is %d\n", r2pipe->child);
-		if (read (r2pipe->output[0], &ch, 1) != 1) {
+		if (read (r2p->output[0], &ch, 1) != 1) {
 			eprintf ("Failed to read 1 byte\n");
-			r2pipe_close (r2pipe);
+			r2pipe_close (r2p);
 			return NULL;
 		}
-		if (ch) {
+		if (ch == -1) {
 			eprintf ("[+] r2pipeipe-io link failed. Expected two null bytes.\n");
-			r2pipe_close (r2pipe);
+			r2pipe_close (r2p);
 			return NULL;
 		}
 		// Close parent's end of pipes
-		close (r2pipe->input[0]);
-		close (r2pipe->output[1]);
-		r2pipe->input[0] = -1;
-		r2pipe->output[1] = -1;
+		close (r2p->input[0]);
+		close (r2p->output[1]);
+		r2p->input[0] = -1;
+		r2p->output[1] = -1;
 	} else {
 		int rc = 0;
 		if (cmd && *cmd) {
 			close (0);
 			close (1);
-			dup2 (r2pipe->input[0], 0);
-			dup2 (r2pipe->output[1], 1);
-			close (r2pipe->input[1]);
-			close (r2pipe->output[0]);
-			r2pipe->input[1] = -1;
-			r2pipe->output[0] = -1;
-			rc = r_sandbox_system (cmd, 0);
+			dup2 (r2p->input[0], 0);
+			dup2 (r2p->output[1], 1);
+			close (r2p->input[1]);
+			close (r2p->output[0]);
+			r2p->input[1] = -1;
+			r2p->output[0] = -1;
+			rc = r_sandbox_system (cmd, 1);
+			fprintf (stderr, "return code %d for %s\n", rc, cmd);
+			fflush (stderr);
+			// trigger the blocking read
```

Comment:
what does this code actually do? what does it solve? how?

Reply:
if r2 is not in PATH or not available this unblocks the blocking read above by sending an 0xff

Reply:
Ah, I see now. What if r2 is in PATH instead? Who would send the 0xff? Anyway, I think a nicer comment explaining this thing would be welcome, because it is not immediate to see this, as in the code above you check for -1, here instead you write 0xff.

Reply:
the r2 protocol starts with 2 null bytes. so when spawning r2 with -0 it will write that and pass. it's used as magic to detect that it is an r2pipe peer who's listening in the other side.

Reply:
i added a test